### DISCUSSION TITLE
setting a "min" error on a Field Array via useEffect, but only after initial render

### DISCUSSION THREAD

**Discussion by miller-productions** [react-hook-form/react-hook-form#3733] [Q&A] [answered]
Hi there,

I am working with useFieldArray.

I would like to set a "min" error on the field array, i.e. if the field array is empty.

The error is to be hidden/not shown on initial render. The error should be displayed when the form has been submitted.

-------------

To learn how to do this, I have been reading these two examples posted by Bill...

https://codesandbox.io/s/react-hook-form-fieldsarray-yup-validation-min-length-forked-kb90b?file=/src/index.js

https://github.com/react-hook-form/react-hook-form/blob/master/examples/fieldArrayMinLength.tsx

I also had a look at this related discussion #2605.

--------

In my first attempt, I set the error inside useEffect as follows:

```
  React.useEffect(() => {
    if (!fields.length) {
      setError("myFieldArray", {
        type: "min",
        message: "At least one field is required",
      });
    } else {
      clearErrors("myFieldArray");
    }
  }, [fields, setError, clearErrors, isDirty]);
```

Then, down in the ui I display the error as follows...

```
{errors.myFieldArray && (
  <span>
    At least one item is required
  </span>
)}
```

The problem I have with this approach is that `errors.myFieldArray` will be set on initial render and thus the error message will shown straight away.

I would prefer to not show the error on initial render, but wait until after the form has been submitted (i.e. after a submit button click)

So, I then investigated the use of an initial render flag as follows:

```
const isInitalRender = React.useRef(true);

React.useEffect(() => {
  if (!fields.length && !isInitalRender.current) {
    setError("myFieldArray", {
      type: "min",
      message: "At least one field is required",
    });
  } else {
    clearErrors("myFieldArray");
  }
  
  if (isInitalRender.current) {
    isInitalRender.current = false;
  }
}, [fields, setError, clearErrors, isDirty]);
```

This heads me in sort of the right direction but doesn't work as I had thought. With this approach the error is not set on first render (which is the goal), but it is also not set if you immediately submit the form (i.e. click the submit button) after initial render.

By the looks of it, my confusion has something to do with how I am working with useEffect.

Perhaps there is something else I should be including in the second argument to useEffect... i.e. should I be monitoring something else in addition to the following `[fields, setError, clearErrors, isDirty]` 

Would love to hear someone's thoughts on this.

**Comment by miller-productions** [comment, marked as answer]
This appears to be a potential solution...

Replace the second argument to `useEffect` with this... `[fields, isSubmitting]`

In other words, monitor for changes in `fields` and also `isSubmitting`. When either of those happen, then run the logic check to ensure `myFieldArray` has at least one item, error if it doesn't.

**Reply by miller-productions** [reply]
Further investigation suggests using `isSubmitted` might be better to use here instead of `isSubmitting`

**Reply by Axxxx0n** [reply]
but if we use `isSubmitted` and the only error is in this field array, form will submit before we set error, right?

**Comment by bluebill1049** [comment]
hey @miller-productions have you tried to set `shouldUnregister: false` which will produce an empty array during submission.

**Reply by miller-productions** [reply]
Interesting, I just now read about `shouldUnregister`. I'm not yet familiar enough with it to understand how it relates to this scenario. 

From what I've read, setting `shouldUnregister: false` would mean that data from inputs in the form remains available even after they are unmounted/removed.

In my scenario, unmounting/removing isn't really the issue.

The scenario goes like this...

Goal = don't submit form if field array is empty, but also, don't show error initially on first render of form

1. initialise empty field array, i.e. [ ] 
2. render form
3. don't display any error yet
4. If user clicks submit button, then display error now
5. Of course, if user adds to the field array at any time, then clear the error

It does indeed work with the following approach...

```
  // flag initial render of form
  const isInitalRender = React.useRef(true);

  // validation check for at least one item
  React.useEffect(() => {
    if (!fields.length && !isInitalRender.current) {
      setError("myFieldArray", {
        type: "min",
        message: "At least one item is required",
      });
    } else {
      clearErrors("myFieldArray");
    }
    if (isInitalRender.current) {
      isInitalRender.current = false;
    }
  }, [fields, isSubmitted]);
```

I just wasn't sure if that seemed like the sort of approach others more familiar with `react-hook-form` would use (i.e. since I'm quite new to it).

**Reply by bluebill1049** [reply]
I thought your issue was due to schema not recognize the empty array `[]`, because the thing about Field Array (controlled inputs) when all inputs are not present, the state will show empty `[]` while at React Hook Form because there are no inputs, we just show empty `{}` which is potentially make schema validation harder. I am not an expert on Yup, wondering if there is a way to check if empty key for the FieldArray to throw an error.

eg:

```
data: {} // should display error, Field Array length should be at least 1
```

which right now only works 

```
data: {
  test: []
}
```

**Comment by evangrim** [comment]
@bluebill1049 Thank you for putting the effort into making this work without useEffect :)

For posterity, here's an almost working proof of concept for multiple useFieldArray (there's just one here but you can make another by copy/paste) and error messages for both repeater contents and the useFieldArray as a group, as well as error messages for arbitrary other registered inputs. Hopefully useful to someone, took me a while to figure out :)

Note that ErrorMessage is a very simple custom component here that outputs a styled message if present.

//"./RepeaterContact.js";

import { useForm, useFormState, useWatch } from "react-hook-form";
import ErrorMessage from "./ErrorMessage";

export default function RepeaterContact ({ update, index, value, control, list }) {

    const Display = ({ control, index }) => {
            const data = useWatch({
                control,
                name: `${list}.${index}`
            });

        if (!data?.firstName) return null;
    
        return (
        <div>
            <h3>Saved Dishes</h3>
            <p>
                {data?.firstName} {data?.lastName} {data?.phone}
            </p>
            {/* <p>{Boolean(data?.working) && data?.working && "I am working."}</p> */}
        </div>
        );
    };

    const { register, handleSubmit, formState: {errors} } = useForm({
        defaultValues: value
    });

    return (
    <div>
        <Display control={control} index={index} />

        <input
            placeholder="first name"
            {...register(`firstName`, { required: "Required" })}
        />

        <ErrorMessage message={errors?.firstName?.message}></ErrorMessage>
        <input
            placeholder="last name"
            {...register(`lastName`, { required: "Required" })}
        />
        <ErrorMessage message={errors?.lastName?.message}></ErrorMessage>

        <input
            type="tel"
            placeholder="206 555 1212"
            {...register(`phone`, { required: "Required" })}
        />
        <ErrorMessage message={errors?.phone?.message}></ErrorMessage>

        <button
            type="button"
            onClick={handleSubmit((data) => {
                update(index, data);
            })}
        > Save Contact
        </button>
    </div>
    );
};

---
//app or component in app
const repeaterContacts = "contactsArray";
export default function Application() {

import RepeaterContact from "./RepeaterContact";
import { useForm, useFieldArray, useWatch } from "react-hook-form";
    const { 
        register, 
        handleSubmit, 
        control,
        watch, 
        formState: { errors } 
    } = useForm ();

    const { 
            fields: contactFields,
            append: appendContact,
            update: updateContact, 
            remove: removeContact
        } = useFieldArray({
            control,
            name: repeaterContacts,
            defaultValues: {
                [repeaterContacts]: []
            }
        });
        
    const watchContacts = watch(repeaterContacts);
    console.log('contacts', watchContacts);

        <form onSubmit={handleSubmit(onSubmit)}>
            <h2>Contacts</h2>
            {contactFields.map((field, index) => (
                <fieldset key={field.id}>
                    <RepeaterContact
                        control={control}
                        update={updateContact}
                        index={index}
                        value={field}
                        list={repeaterContacts}
                    />
                    <button
                        className="remove"
                        type="button"
                        onClick={() => removeContact(index)}
                    >
                        Remove
                </button>
                </fieldset>
            ))}

            <button
                type="button"
                onClick={() => {
                    appendContact({
                        firstName: "",
                        lastName: ""
                    });
                }}
            >
                Add Contact
            </button>
        
             <ErrorMessage message={errors?.contactsArray?.root?.message}></ErrorMessage>
       
       
                   <label htmlFor="">
                Test Input<br/>
                <input
                    {...register( "test", { 
                        required: "This is required", 
                        minLength: { value: 4, message: "Minimum length 4 characters" } 
                    })}
                    type="text"
                />
            </label>
            <ErrorMessage message={errors.test?.message}></ErrorMessage>

            <p><input type="submit" /></p>

        </form>
    </div>
    );
}
